proposal/iblis17/README.rst

```diff
@@ -0,0 +1,418 @@
++-------------+--------------+------------+--------------+---------------------------------------------------------------------+
+
+
+Problem to solve
+===============================================================================
+
+There are demands of table-like data structures that can hold sorted time axis,
+heterogeneous type and dynamic length time series data.
+This proposal proposes to build a series of data type, interface and operations
+that can fulfill the demands.
+
+
+System architecture
+===============================================================================
+
+Time Type
+----------------------------------------------------------------------
+
+The supported time types can be used as elements of time axis.
+
+#. ``Dates.TimeType``, the abstract time type from Julia's stdlib.
+#. ``Dates.Period``, the abstract period type from Julia's stdlib.
+   It is used for a time interval or a time offset relative to some reference
+   point.
+
+
+Vector of Timestamps
+----------------------------------------------------------------------
+
+A vector of timestamps should be sorted, in order to apply time-related operations.
+Allowing repeating timestamp or not is one of the key properties.
+In this proposal, repeating timestamp in a vector should be allowed.
+In some real cases, sensor generated timestamps might be the same due to the
+inaccurate timer or time resolution,
+but allow repeating does introduce some API design issues.
+
+In case of repeated timestamps, it can be solved by introduce the secondary index,
+assume that the order from user input is in ascending order.
+The secondary index can be automatically built. For instance:
+
++----------+-----------------+--------+
+| Timestap | Secondary Index | Valueu |
++==========+=================+========+
+| 10:00:00 | 1               | foo    |
++----------+-----------------+--------+
+| 10:00:00 | 2               | bar    |
++----------+-----------------+--------+
+| 10:00:00 | 3               | baz    |
++----------+-----------------+--------+
+| 10:30:00 | 1               | qaz    |
++----------+-----------------+--------+
+| ...      |                 |        |
++----------+-----------------+--------+
+
+The we can provides two sets of APIs.
+
+#. All the APIs in the first set will take ``secondary index = 1`` and compute.
+   This set is designed for the case that there aren't repeating in the timestamps,
+   or user want to ignore the repeating in specific operations.
+
+#. The second set can accept the secondary index as a argument, and the
+   return type will be different from the first set of APIs. Most of APIs in this
+   set should return the table type that same or similar to the input table.
+
+
+
+API descriptions
+===============================================================================
+
+There is a abstract type ``AbstractTimeSeries`` in this project,
+all table-like data structure should be a subtype of it.
+
+``AbstractTimeSeries`` Design
```

Comment:
The API descriptions are very complete.  It may be too complete as a proposal, but it does more good than harm.  I guess you might already have a prototype working?  Regardless, I think the descriptions here will help your implementation.

It may need to be refined with your implementation iterations, still.